### ISSUE TITLE
CSS Custom Highlight API Module Level 1

### ISSUE THREAD

**Issue by frivoal** [w3ctag/design-reviews#584]
你好 TAG! 吃饭了吗？

I'm requesting a TAG review of CSS Custom Highlight API Module Level 1, which was recently published as a FPWD.

The Custom Highlight API extends the concept of highlight pseudo-elements (see CSS Pseudo-Elements 4 §3 Highlight Pseudo-elements) by providing a way for web developers to style the text of arbitrary Range objects, rather than being limited to the user agent defined ::selection, ::spelling-error, and ::grammar-error. This is useful in a variety of scenarios, including editing frameworks that wish to implement their own selection, find-on-page over virtualized documents, multiple selection to represent online collaboration, or spellchecking frameworks. The Custom Highlight API provides a programmatic way of adding and removing highlights that do not affect the underlying DOM structure, but instead applies styles to text based on range objects, accessed via the ::highlight() pseudo element.

  - Explainer¹ (minimally containing user needs and example code): https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/highlight/explainer.md (largely superseded by the spec's own intro section, but may still be of interest for discussions of alternatives)
  - Specification URL: https://www.w3.org/TR/css-highlight-api-1/
  - Tests: none yet :(
  - Security and Privacy self-review²: https://lists.w3.org/Archives/Public/www-style/2020Dec/0007.html
  - GitHub repo (if you prefer feedback filed there): https://github.com/w3c/csswg-drafts (one issue per comment please)
  - Primary contacts (and their relationship to the specification):
      - Florian Rivoal (@frivoal), Invited Expert, co-editor
      - Sanket Joshi (@sanketj), Microsoft, co-editor
      - Megan Gardner, Apple, co-editor
      - Rossen Atanassof, @atanassov, Microsoft, co-chair
      - Alan Stearns @astearns, Adobe, co-chair
  - Organization(s)/project(s) driving the specification: CSSWG
  - Key pieces of existing multi-stakeholder review or discussion of this specification:
    * https://github.com/MicrosoftEdge/MSEdgeExplainers/issues?q=label%3A%22Highlight+API%22+
    * https://github.com/w3c/csswg-drafts/issues?q=is%3Aissue+label%3Acss-highlight-api-1+
  - External status/issue trackers for this specification (publicly visible, e.g. Chrome Status):
    * Implemented in Safari [since release preview 99](https://developer.apple.com/safari/technology-preview/release-notes/#r99)

Further details:

  - [X] I have reviewed the TAG's [API Design Principles](https://w3ctag.github.io/design-principles/)
  - Relevant time constraints or deadlines: Nothing in particular
  - The group where the work on this specification is currently being done: CSSWG
  - The group where standardization of this work is intended to be done (if current group is a community group or other incubation venue): CSSWG
  - Major unresolved issues with or opposition to this specification: No known opposition, event part still missing
  - This work is being funded by: Bloomberg, Microsoft, Apple

We'd prefer the TAG provide feedback as 🐛 open issues in our GitHub repo for **each point of feedback**

**Comment by LeaVerou** [comment]
To everyone else reviewing: The differences between the explainer and the actual spec are at this point **very substantial**. I would suggest that people review the [spec's own intro section](https://drafts.csswg.org/css-highlight-api-1/#intro) instead of the explainer to get an overview of the current status.

Yes, it is mentioned above in passing, but thought I'd highlight it (pun intended 😁) in case others also start from the explainer by force of habit despite @florian's comment, and then realize that all their comments do not apply to the current spec anymore.

**Comment by florian** [comment]
Lea, I'm pretty sure you meant to @ mention some other Florian.

**Comment by LeaVerou** [comment]
@florian Oops, fixed now, thanks!

**Comment by LeaVerou** [comment]
@hober and I discussed this today in our VF2F. We noticed that `HighlightsRegister` is a setlike, but throws when a `Highlight` object with an existing registered name, which indicates that perhaps maplike would have been more appropriate? It's not ideal developer experience if developers need to iterate the entire data structure before registering their object, to check if an existing one of the same name exists. Maplike would also provide an easy way to get all ranges under a given name.
